core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -86,14 +89,74 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
   }
 
   /**
-   * Registers a given broker in zookeeper as the controller.
+   * Registers a given broker in zookeeper as the controller and increments controller epoch.
+   * @return the (updated controller epoch, epoch zkVersion) tuple
    * @param controllerId the id of the broker that is to be registered as the controller.
-   * @param timestamp the timestamp of the controller election.
-   * @throws KeeperException if an error is returned by ZooKeeper.
-   */
-  def registerController(controllerId: Int, timestamp: Long): Unit = {
-    val path = ControllerZNode.path
-    checkedEphemeralCreate(path, ControllerZNode.encode(controllerId, timestamp))
+   * @throws ControllerMovedException if fail to create /controller or fail to increment controller epoch.
+   */
+  def registerControllerAndIncrementControllerEpoch(controllerId: Int): (Int, Int) = {
+    val timestamp = time.milliseconds()
+
+    // Read /controller_epoch to get the current controller epoch and zkVersion,
+    // create /controller_epoch with initial value if not exists
+    val (curEpoch, curEpochZkVersion) = getControllerEpoch
+      .map(e => (e._1, e._2.getVersion))
+      .getOrElse(maybeCreateControllerEpochZNode())
+
+    // Create /controller and update /controller_epoch atomically
+    val newControllerEpoch = curEpoch + 1
+    val expectedControllerEpochZkVersion = curEpochZkVersion
+
+    debug(s"Try to create ${ControllerZNode.path} and increment controller epoch to $newControllerEpoch with expected controller epoch zkVersion $expectedControllerEpochZkVersion")
+
+    def checkControllerAndEpoch(): (Int, Int) = {
+      val curControllerId = getControllerId.getOrElse(throw new ControllerMovedException(
+        s"The ephemeral node at ${ControllerZNode.path} went away while checking whether the controller election succeeds. " +
+          s"Aborting controller startup procedure"))
+      if (controllerId == curControllerId) {
+        val (epoch, stat)  = getControllerEpoch.getOrElse(throw new ControllerMovedException(
+          "Controller moved to another broker. Aborting controller startup procedure"))
+        // If the epoch is the same as newControllerEpoch, it is safe to infer that the returned epoch zkVersion
+        // is associated with the current broker during controller election because we already knew that the zk
+        // transaction succeeds based on the controller znode verification. Other rounds of controller
+        // election will result in larger epoch number written in zk.
+        if (epoch == newControllerEpoch)
+          return (newControllerEpoch, stat.getVersion)
```

Comment:
I have two questions here:

1) When would we enter a scenario that `controllerId == curControllerId` and `epoch != newControllerEpoch)`?

2) Currently when this happens, `checkControllerAndEpoch` will throw `ControllerMovedException()`, which is caught in `KafkaController.elect()` and trigger `maybeResign()`. However `maybeResign()` will do nothing because `controllerId == curControllerId` and this broker is considered to be the active controller. In this case no other broker will be controller. And the current broker will not function properly as controller because it has not executed `onControllerFailover()`.

So maybe we should throw `IllegalStateException` here if `controllerId == curControllerId` and `epoch != newControllerEpoch)`?

Reply:
1. Because we first get /controller and then get /controller_epoch (rather than do it atomically), it is possible that after we see `controllerId == curControllerId` and before we get /controller_epoch, another round of controller election is triggered. In this case, we will see `epoch != newControllerEpoch`.
2. When we see `epoch != newControllerEpoch`, `controllerId == curControllerId` does not hold because another broker must become the controller. In this case, `maybeResign` will work fine.

Reply:
Thanks for the explanation. This makes sense.

Reply:
It is safe to replace the /controller payload check with session id check but we still need to read /controller_epoch to get back the corresponding zk version if we loss connection when doing the zk transaction.

Reply:
My understanding is that the approach using ephemeral owner has the same performance and correctness guarantee as the current approach which uses epoch from the znode data. 

The approach using ephemeral owner is probably more intuitive/readable because it exactly  handles the root cause of `NodeExistsException | _: BadVersionException`, i.e. `checkControllerAndEpoch()` should effectively translate NodeExistsException/BadVersionException to `Code.OK` if and only if `ephemeral owner of the controller path equals to the ZK session id`. And it is also more consistent with the existing logic in `checkedEphemeralCreate()`.

@hzxa21 If it sounds reasonable, maybe we can have a minor followup patch (without requiring a JIRA ticket) to improve it.

Reply:
I was thinking that the ephemeral owner approach will be cheaper since we only need to read the controller path, not both controller path and the controller epoch path.

Reply:
@junrao From the correctness and performance point of view, checking `/controller` payload and checking `/controller` ephemeral owner is the same. The question is whether we need to read `/controller_epoch`. The reason why I do it is because if `NodeExistsException| BadVersionException` happens, we no longer have the `Stat` (new zkVersion) of `/controller_epoch` even though the `/controller_epoch` update succeeds in zookeeper server. We can avoid the extra read on `/controller_epoch` if we can assume that zkVersion is always incremented by one. Since `/controller_epoch` zkVersion is critical for us after this patch and zookeeper doc does not explicitly say that this assumption holds, I think it is safer to do one extra read during controller election.

Reply:
@hzxa21 : That's a great point. Thanks. Then we can just keep the code as it is.